src/format/packed/mod.rs

```diff
@@ -51,18 +42,72 @@ pub enum Variant {
 #[cfg(debug_assertions)]
 impl<T: Element> ::format::Validate for Packed<T> {
     fn validate(&self) {
-        assert_eq!(self.values.len(), storage!(self.size));
+        assert_eq!(self.values.len(), triangular_elems(self.size));
     }
 }
 
 size!(Packed, size, size);
 
+/// Computes the offset into the array for the upper triangular representation
+/// Copied from http://www.netlib.org/lapack/lug/node123.html
+#[inline]
+pub fn upper_triangular_offset(n: usize, i: usize, j: usize) -> usize {
+    debug_assert!(i <= j);
+    debug_assert!(i <= n);
+    debug_assert!(j <= n);
+    // Source material uses 1-based indexing
+    let i = i + 1;
+    let j = j + 1;
+    i + j * (j - 1) / 2 - 1
+}
+
+/// Computes the offset into the array for the lower triangular representation
+/// Copied from http://www.netlib.org/lapack/lug/node123.html
+#[inline]
+pub fn lower_triangular_offset(n: usize, i: usize, j: usize) -> usize {
+    debug_assert!(j <= i);
+    debug_assert!(i <= n);
+    debug_assert!(j <= n);
+    // Source material uses 1-based indexing
+    let i = i + 1;
+    let j = j + 1;
+    i + (2 * n - j) * (j - 1) / 2 - 1
+}
+
+/// Computes how many unique elements are in a symmetric square matrix of size nxn
+#[inline]
+pub fn triangular_elems(n: usize) -> usize {
+    n * (n + 1) / 2
+}
+
 impl<T: Element> Packed<T> {
     /// Create a zero matrix.
     pub fn new<S: Size>(size: S, variant: Variant) -> Self {
         let (rows, _columns) = size.dimensions();
         debug_assert!(rows == _columns);
-        new!(rows, variant, vec![T::zero(); storage!(rows)])
+        new!(rows, variant, vec![T::zero(); triangular_elems(rows)])
+    }
+
+    /// Read an element from the matrix
+    pub fn get<P: Position>(&self, position: P) -> Option<&T> {
```

Comment:
The benefit to returning an `Option<&T>` is twofold: first, if the position is invalid, it's bad form to panic in a library, so this allows us to communicate the problem without crashing the system. The other benefit is that `Option<&T>` is specially optimized so that `None` is represented as the null pointer (i.e. `0`), which means this is very efficient.

In the issue related to indexing, I'd like to work out the general API solution, since it seems to me that indexing and natural-order iteration are related.